### DISCUSSION TITLE
Restoring a subscription after the server connection is lost

### DISCUSSION THREAD

**Discussion by shayakhmetovi** [FreeOpcUa/python-opcua#1534] [Q&A] [answered]
I use client [python-opcua](https://github.com/FreeOpcUa/python-opcua) to connect to server [node-opcua](https://github.com/node-opcua/node-opcua). The connection status is monitored by keepalive. Created a subscription for 10-20 nodes.
Is it possible to check the subscription status: is it working or not? or how to catch that the subscription is no longer active? Is there any mechanism to [restore the subscription](https://reference.opcfoundation.org/Core/Part4/v104/docs/5.13.7)? For example, the server disconnects and the subscription needs to be restored

**Comment by schroeder-** [comment, marked as answer]
Switch to asyncua first. 
Then you can used a modified https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client-reconnect.py.

If you add/remove nodes on demand, you have to remember the subscribed nodes your self, for example make a list or dict and store all subscribed nodes. When you reconnect you subscribe to all nodes in this list.
